core/playlistmanager.go

```
@@ -0,0 +1,262 @@
+	// InsertLink link to segment that already in the storage
+	// (Transcoder wrote segment in B's owned storage and returned link, B adds this link to video source)
+	// returns true if it was acutally inserted
+	InsertLink(turi *net.TypedURI, duration float64) (bool, error)
+
+	GetHLSMasterPlaylist() *m3u8.MasterPlaylist
+
+	GetHLSMediaPlaylist(streamID StreamID) *m3u8.MediaPlaylist
+
+	GetSession() drivers.OSSession
+	// Ends object storage's session
+	EndSession()
+}
+
+type BasicPlaylistManager struct {
+	storageSession drivers.OSSession
+	manifestID     ManifestID
+	// Live playlist used for broadcasting
+	masterLivePList *m3u8.MasterPlaylist
+	// Full playlist is stored into OS for later reference
+	masterFullPList *m3u8.MasterPlaylist
+	mediaLiveLists  map[StreamID]*m3u8.MediaPlaylist
+	mediaFullLists  map[StreamID]*m3u8.MediaPlaylist
+	mapSync         *sync.Mutex
+}
+
+// NewBasicPlaylistManager create new BasicPlaylistManager struct
+func NewBasicPlaylistManager(manifestID ManifestID,
+	storageSession drivers.OSSession) *BasicPlaylistManager {
+
+	bplm := &BasicPlaylistManager{
+		storageSession:  storageSession,
+		manifestID:      manifestID,
+		masterLivePList: m3u8.NewMasterPlaylist(),
+		masterFullPList: m3u8.NewMasterPlaylist(),
+		mediaLiveLists:  make(map[StreamID]*m3u8.MediaPlaylist),
+		mediaFullLists:  make(map[StreamID]*m3u8.MediaPlaylist),
+		mapSync:         &sync.Mutex{},
+	}
+	return bplm
+}
+
+func (mgr *BasicPlaylistManager) ManifestID() ManifestID {
+	return mgr.manifestID
+}
+
+func (mgr *BasicPlaylistManager) GetSession() drivers.OSSession {
+	return mgr.storageSession
+}
+
+// InsertHLSSegment ...
+func (mgr *BasicPlaylistManager) InsertHLSSegment(streamID StreamID,
+	seg *stream.HLSSegment) (*net.TypedURI, error) {
+
+	if err := mgr.isRightStream(streamID); err != nil {
+		return nil, err
+	}
+	sid := string(streamID)
+	fmpl := mgr.getFullPL(streamID)
+	if fmpl == nil {
+		panic("no full media pl")
```

Is this really a fatal error? Why panic rather than log and return?

---

it was for debugging, missed when was cleaning up